client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt

```diff
@@ -0,0 +1,237 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.shareaware.client
+
+import com.google.protobuf.Message
+import io.grpc.ManagedChannelBuilder
+import io.spine.base.CommandMessage
+import io.spine.base.EntityState
+import io.spine.base.EventMessage
+import io.spine.client.Client
+import io.spine.client.ClientRequest
+import io.spine.client.EventFilter
+import io.spine.core.UserId
+import io.spine.examples.shareaware.WalletId
+import io.spine.examples.shareaware.wallet.command.CreateWallet
+import io.spine.examples.shareaware.wallet.event.WalletCreated
+import io.spine.util.Exceptions
+import java.util.*
+import java.util.concurrent.CompletableFuture
+
+/**
+ * Interacts with the app server via gRPC.
+ */
+public class DesktopClient private constructor(
+    host: String,
+    port: Int
+) {
+
+    public companion object {
+
+        @Volatile
+        private var instance: DesktopClient? = null
+
+        public fun init(host: String, port: Int): DesktopClient =
+            instance ?: synchronized(this) {
+                instance ?: DesktopClient(host, port).also { instance = it }
+            }
+    }
+
+    private val client: Client
+    private lateinit var user: UserId
+    private lateinit var walletId: WalletId
+
+    init {
+        val channel = ManagedChannelBuilder
+            .forAddress(host, port)
+            .usePlaintext()
+            .build()
+        client = Client
+            .usingChannel(channel)
+            .build()
+    }
+
+    /**
+     * Sends command to authenticate user in the system.
+     *
+     * A command `CreateWallet` is used for user creation because in the ShareAware
+     * there is no functional to register or authenticate the user, yet.
+     */
+    public fun authenticateUser() {
+        val userId = UUID.randomUUID().toUserId()
+        val walletId = userId.toWalletId()
+        val walletField = WalletCreated.Field.wallet()
+        val walletCreated: CompletableFuture<WalletCreated> = CompletableFuture()
+        this.subscribeToEvent(
+            WalletCreated::class.java,
+            EventFilter.eq(walletField, walletId)
+        )
+        {
+            walletCreated.complete(it)
+        }
+        command(createWallet(walletId))
+        if (!::user.isInitialized) {
```

Comment:
I am not sure why we use such a syntax here. Why not just call the method?

Same below.

Reply:
I have defined this variable with `lateinit` property to avoid null instantiation. As Kotlin documentation says any call of this variable before its initialization will raise an exception. To find out if this variable is initialized or not we need to use property reference. 
<a href="https://kotlinlang.org/docs/whatsnew12.html#check-whether-a-lateinit-var-is-initialized:~:text=on%20v.1.8.21-,Check%20whether%20a%20lateinit%20var%20is%20initialized,-You%20can%20now">Documentation here</a>

Reply:
Ideally you don't need to check for that at all. Maybe, just once, somewhere during the application bootstrap.

The issue here is that `user` is never explicitly initialised in scope of this method prior to this check. Or it is, but I cannot see that.

How about we authenticate user right when the instance of `Client` is created? And then continue our journey knowing that `user` is always present?